stdlib/public/stubs/SwiftNativeNSXXXBase.mm.gyb

```diff
@@ -105,6 +105,22 @@ swift_stdlib_NSObject_isEqual(NSObject *NS_RELEASES_ARGUMENT lhs,
   return Result;
 }
 
+// _direct type metadata for Swift._NSContiguousString
+SWIFT_RUNTIME_STDLIB_INTERFACE
+extern "C" ClassMetadata _TMCs19_NSContiguousString;
+
+extern "C" {
+  struct _SwiftStackAllocatedNSContiguousString {
+    // Heap object header.
+    struct HeapObject header;
+
+    // _StringCore.
+    void *baseAddress;
+    uintptr_t countAndFlags;
+    void *owner;
+  };
```

Comment:
I think this is promising.  You could in fact add a `withNSConstantString() { ... }` API on `StringCore` that would encapsulate creating the stack allocated struct.  I think this would cut down the boilerplate a lot and would address most of the concerns.